kernel/arch/dreamcast/hardware/sq.c

```diff
@@ -115,38 +122,48 @@ void sq_wait(void) {
 
 /* Copies n bytes from src to dest, dest must be 32-byte aligned */
 __attribute__((noinline)) void *sq_cpy(void *dest, const void *src, size_t n) {
-    uint32_t *d = SQ_MASK_DEST(dest);
     const uint32_t *s = src;
+    void *curr_dest = dest;
+    uint32_t *d;
+    size_t nb;
 
     /* Fill/write queues as many times necessary */
     n >>= 5;
 
-    /* Exit early if we dont have enough data to copy */
-    if(n == 0)
-        return dest;
-
-    sq_lock(dest);
-
-    /* If src is not 8-byte aligned, slow path */
-    if ((uintptr_t)src & 7) {
-        while(n--) {
-            dcache_pref_block(s + 8); /* Prefetch 32 bytes for next loop */
-            d[0] = *(s++);
-            d[1] = *(s++);
-            d[2] = *(s++);
-            d[3] = *(s++);
-            d[4] = *(s++);
-            d[5] = *(s++);
-            d[6] = *(s++);
-            d[7] = *(s++);
-            sq_flush(d);
-            d += 8;
+    while (n > 0) {
+        /* Transfer maximum 512 KiB at once. This is because when using the
+         * MMU the SQ area is a 1 MiB page, and the destination address may
+         * not be on a page boundary. */
+        nb = n > 0x4000 ? 0x4000 : n;
```

Comment:
Well there's a comment right above explaining it :)
The SQ area with MMU is 1MiB, so to avoid getting a segmentation fault, we move the SQ window after each 512 KiB transferred (and not 1 MiB because the destination address may start from an offset within the SQ page).

Reply:
So the offset cannot be more than 512KiB? My question isn't the reason for limiting to less than the size of the page, which is explained in the comment, but why the specific amount 512KiB is used.

Reply:
I guess I was confused :)
It's 1 MiB now.